### PR TITLE
Drag'n'Drop support

### PR DESCRIPTION
**Pull request by StCyr**
Drag'n'Drop support implementation with https://github.com/Vivify-Ideas/vue-draggable

Remaining issues:

- [ ] It seems the code isn't properly initialized: If you load the mail apps from your browser, the drag'n'drop functionality won't work directly. You first have to perform an action (eg: selecting another email than the default one displayed) before the drag'n'drop functionality will work;
- [ ] Shouldn't navigate when the moved message is not tghe one currently displayed
- [ ] Do not expand when folder's account is not the same as envelope account
- [ ] Polishing

### CODE REVIEW

**Comment by StCyr** [comment]
well, fixing support for the `onDragstart` and `onDragend` functions was quite easy :-p

**Comment by StCyr** [comment]
Hi @ChristophWurst,

Some feedback would be welcome.

In particular, I'm planing to re-use the MessagesController's move function. Could you confirm it's ok to re-use it for this purpose?

Br,

**Comment by StCyr** [comment]
Hi @ChristophWurst 

So, I've added a new commit so that the proper api endpoint gets called (MesssagesController's move function) and... it works! \o/\o/\o/ !

The following issues are still current:

1. The EnvelopeList isn't refreshed to show that the message has been moved;
2. It seems the code isn't properly initialized: If you load the mail apps from your browser, the drag'n'drop functionality won't work directly. You first have to perform an action (eg: selecting another email than the default one displayed) before the drag'n'drop functionality will work.

Also, I don't issue store commit commands yet, while I should probably do (it might even automagicaly solve the EnvelopeList refresh issue. Who knows? :-) ).

Feedback would be much appreciated :-)

Cyrille

**Comment by StCyr** [comment]
Commited a change to remove envelope and message after move.

Update on current issues:
1. Drag'n'Drop doesn't work from unified inbox;
2. It seems the code isn't properly initialized: If you load the mail apps from your browser, the drag'n'drop functionality won't work directly. You first have to perform an action (eg: selecting another email than the default one displayed) before the drag'n'drop functionality will work;
3. In case of error when moving the message on the IMAP server, it's highly possible that the envelope won't be added back properly because of a shortcoming in my code.

**Comment by ChristophWurst** [comment]
> In particular, I'm planing to re-use the MessagesController's move function. Could you confirm it's ok to re-use it for this purpose?

Yes! That's what that method was added for :)



>     1. The EnvelopeList isn't refreshed to show that the message has been moved;

The change has to go through the Vuex store. Then it will update automatically.

> Also, I don't issue store commit commands yet, while I should probably do (it might even automagicaly solve the EnvelopeList refresh issue. Who knows? :-) ).

It will :rocket: 

>     1. Drag'n'Drop doesn't work from unified inbox;

Fine by me. We also had this limitation on the Backbone version of this app's front-end.

>     3\. In case of error when moving the message on the IMAP server, it's highly possible that the envelope won't be added back properly because of a shortcoming in my code.

Again, if the store is used (correctly :wink:) this won't be a problem.

File: src/router.js [hunk 1]
```diff
@@ -8,6 +9,7 @@ const KeyboardShortcuts = () => import('./views/KeyboardShortcuts')
 const Setup = () => import('./views/Setup')
 
 Vue.use(Router)
+Vue.use(VueDraggable)
```
[the remark this answers is not in the document]
Comment (StCyr) [reply src/router.js]: I've moved it in `main.js` but I don't think I know enough about js and the frameworks used here to move it somewhere else

File: src/views/Home.vue [hunk 2]
```diff
@@ -1,5 +1,36 @@
...
+	<Content v-shortkey.once="['c']" app-name="mail" @shortkey.native="onNewMessage"
+		v-drag-and-drop:options="{
+                       	dropzoneSelector: '.icon-folder',
+                       	draggableSelector: '.app-content-list-item',
+                       	onDragstart: function(e){ debug('start dragging') },
+                       	onDragend: function(e){
+				debug('stop dragging')
+				startAccount = e.items[0].attributes.href.baseURI.match(/message\/(\d)/)
+				targetAccount = e.droptarget.href.match(/accounts\/(\d)/)
+				if (targetAccount[1] == startAccount[1]) {
```
[the remark this answers is not in the document]
Comment (StCyr) [reply src/views/Home.vue]: ok done

**Comment by StCyr** [comment]
Hi @ChristophWurst,

I've made the changes you've requested.

I hope it's OK that I've moved the `Vue.use(VueDraggable)` statement in `main.js` because I've no idea where to put it elsewhere.

The following issues remain:

1. Drag'n'Drop doesn't work from unified inbox;
  => won't fix probably
2. It seems the code isn't properly initialized: If you load the mail apps from your browser, the drag'n'drop functionality won't work directly. You first have to perform an action (eg: selecting another email than the default one displayed) before the drag'n'drop functionality will work;
  => no idea why
3. In case of error when moving the message on the IMAP server, it's highly possible that the envelope won't be added back properly because of a shortcoming in my code.
  => to do

Other todo's:

1. remove unecessary `console.debug` statements
2. investigate if code should be moved back in `onDrop`

**Comment by StCyr** [comment]
Ok, looks good here. 

The following issues remain:

1. Drag'n'Drop doesn't work from unified inbox;
    => won't fix probably (Though I'll investigate a little because it doesn't really make sense to me)
2. It seems the code isn't properly initialized: If you load the mail apps from your browser, the drag'n'drop functionality won't work directly. You first have to perform an action (eg: selecting another email than the default one displayed) before the drag'n'drop functionality will work;
    => no idea why

Possible improvement:

1. Highlight folders when envelope is dragged over them
  => Investigating

**Comment by StCyr** [comment]
>Highlight folders when envelope is dragged over them

hmmm, it looks like changing `nextcloud-vue's AppNaviationItem` component is needed to do that.

In https://github.com/nextcloud/nextcloud-vue/blob/master/src/components/AppNavigationItem/AppNavigationItem.vue:

```html
<!-- Navigation item -->
	<nav-element v-else :id="item.id" v-bind="navElement(item)"
:title="item.title" :class="[{'icon-loading-small': item.loading, 'open': opened, 'collapsible': collapsible }, item.classes]">
	<!-- Bullet -->
		<div v-if="item.bullet" :style="{ backgroundColor: item.bullet }" class="app-navigation-entry-bullet" />

		<button v-if="collapsible" class="collapse" @click.prevent.stop="toggleCollapse" />

		<a :class="item.icon" :href="(item.href) ? item.href : '#'"
---		    @click="callPreventStop(simpleAction, $event)">
+++		    @click="callPreventStop(simpleAction, $event)"
+++                 @dragenter="highlightMe('true')"
+++                 @dragleave="highlightMe('false')"
+++                 >
			<img v-if="item.iconUrl" :alt="item.text" :src="item.iconUrl">
			{{ item.text }}
</a>
```

**Comment by StCyr** [comment]
Hi,

I've fixed an issue (Drag'n'Drop doesn't work from unified inbox")

The following issue remain:

1. It seems the code isn't properly initialized: If you load the mail apps from your browser, the drag'n'drop functionality won't work directly. You first have to perform an action (eg: selecting another email than the default one displayed) before the drag'n'drop functionality will work;
    => no idea why

Possible improvement:

1. Highlight folders when envelope is dragged over them
    => Looks complicated.  I'm thinking about 3 possible solutions:
    1. updating `AppNavigationItem.vue` (see here above);
    2. updating vueDraggable (https://github.com/Vivify-Ideas/vue-draggable);
    3. using [WAI-ARIA](https://en.wikipedia.org/wiki/WAI-ARIA) as done in https://vivify-ideas.github.io/vue-draggable/example/#start

**Comment by StCyr** [comment]
I think, merging this PR would solve: 

https://github.com/nextcloud/mail/issues/254
https://github.com/nextcloud/mail/issues/659
https://github.com/nextcloud/mail/issues/670
https://github.com/nextcloud/mail/issues/1538
https://github.com/nextcloud/mail/issues/786

**Comment by StCyr** [comment]
Hi @ChristophWurst,

I've improved my PR to fix some failed travis checks. There's only one left failing but it's not caused by my changes I believe.

Could you please take the lead and tell me if there's something else that would be needed to have this PR merged?

For the record, here are the remaining issues I've find:

1. It seems the code isn't properly initialized: If you load the mail apps from your browser, the drag'n'drop functionality won't work directly. You first have to perform an action (eg: selecting another email than the default one displayed) before the drag'n'drop functionality will work;
2. Folders are not highlighted when an envelope is dragged over them

Also, until now I've only tested this PR against Firefox on Windows 10 and MacOS.

Br,

Cyrille

**Comment by StCyr** [comment]
- **Highlighting folders when dragged over**

I'd some successes with [WAI-ARIA](https://en.wikipedia.org/wiki/WAI-ARIA):

```
+/* Highlight folders when dragged over */
.icon-folder[aria-dropeffect="move"]:focus,
.icon-folder[aria-dropeffect="move"].dragover
{
       background-color: rgb(240,240,240);
}
```

However, this is quite limited as it can't implement any logic afaik. And, some logic is needed: For example, folders should automaticaly get developed when dragged over so that subfolders can be shown, and messages dragged into them.

So, I'm really stuck now regarding this PR

**Comment by StCyr** [comment]
I've find a new issue:

* Next/Previous envelope isn't shown when moving the currently displayed envelope.

So, the list of issues is:

1. It seems the code isn't properly initialized: If you load the mail apps from your browser, the drag'n'drop functionality won't work directly. You first have to perform an action (eg: selecting another email than the default one displayed) before the drag'n'drop functionality will work;
2. Folders are not highlighted when an envelope is dragged over them nor automatically expanded.
3. Next/Previous envelope isn't shown when moving the currently displayed envelope.

**Comment by nicolas-san** [comment]
Can I help somehow ? (I'm php dev, I can do js, but not an expert) - I really like have this back :)

**Comment by StCyr** [comment]
You can try for sure. And I will help you.

But I didn’t work on this PR since I wrote it’s blocked; I was hoping some nextcloud expert could tackle the remaining issues :(

**Comment by nicolas-san** [comment]
I'm not familiar with travis, can I fork your repo and try to build on my own, to reproduce the error ?

**Comment by StCyr** [comment]
> 2 - go to the travis-ci site
> 3 - sign up with your github acces

Wow, I've nevet logged in the travis-ci site :-)

**Comment by nicolas-san** [comment]
@StCyr I am not very familiar with this, but can you update your fork, and rebase your branch with the master updated and try to push again, si whats travis react ?
 (If someone knew a better way to do this, go ahead :) )

**Comment by StCyr** [comment]
> @StCyr I am not very familiar with this, but can you update your fork, and rebase your branch with the master updated and try to push again, si whats travis react ?
> (If someone knew a better way to do this, go ahead :) )

I've rebased my branch

**Comment by ChristophWurst** [comment]
> I've rebased my branch

Thanks! Mind squashing your commits into a single one?

**Comment by StCyr** [comment]
commits squashed.

Remaining issues:


- [ ] It seems the code isn't properly initialized: If you load the mail apps from your browser, the drag'n'drop functionality won't work directly. You first have to perform an action (eg: selecting another email than the default one displayed) before the drag'n'drop functionality will work;
- [ ] Folders are not highlighted when an envelope is dragged over them nor automatically expanded.
- [ ] Next/Previous envelope isn't shown when moving the currently displayed envelope.

**Comment by nicolas-san** [comment]
@ChristophWurst  What is the best way to contribute to this PR ?
I have already forked the @StCyr fork, but if I change something, I have to PR his repo, before he can merge and PR here ?
(It seems complicated, and I do not know if a simple flow exist, or if it's managed another way)

**Comment by nicolas-san** [comment]
This version cannot be used with nextcloud 16 ? 

>  This app cannot be installed because the following dependencies are not fulfilled:

>    Server version 17 or higher is required.

**Comment by ChristophWurst** [comment]
> (It seems complicated, and I do not know if a simple flow exist, or if it's managed another way)

Take the branch, add you changes and open a second PR ;)

> This version cannot be used with nextcloud 16 ? This is expected ?

Yes, the master branch just dropped 16 support as 17 is about to be released.

**Comment by nicolas-san** [comment]
> commits squashed.
> 
> Remaining issues:
> 
>     * [ ]  It seems the code isn't properly initialized: If you load the mail apps from your browser, the drag'n'drop functionality won't work directly. You first have to perform an action (eg: selecting another email than the default one displayed) before the drag'n'drop functionality will work;
> 
>     * [ ]  Folders are not highlighted when an envelope is dragged over them nor automatically expanded.
> 
>     * [ ]  Next/Previous envelope isn't shown when moving the currently displayed envelope.

* [ ] when you scroll down, the new message displayed in the envelopeList are not initialized has draggable

**Comment by StCyr** [comment]
>     * when you scroll down, the new message displayed in the envelopeList are not initialized has draggable

Hello, it works properly here

**Comment by StCyr** [comment]
Regarding the fact that vue-draggable doesn't initialize properly, I've observed the following:

When I reload the app (unified inbox view), the elements of my EnvelopeList look like this in the DOM:

image

Then, as soon as I click on another message (it seems that's actually as soon as I click on anything that refreshes the page), they turn to the following:

image

That's, their `draggable` attribute is set to `true`.

Any @nextcloud/mail or @nextcloud/vue expert has an idea why the `draggable` attribute of the `app-content-list-item`'s is not directly set? pinging also @jancborchardt and @ChristophWurst for good measure :-)

**Comment by nicolas-san** [comment]
> Regarding the fact that vue-draggable doesn't initialize properly, I've observed the following:
> 
> When I reload the app (unified inbox view), the elements of my EnvelopeList look like this in the DOM:
> 
> image
> 
> Then, as soon as I click on another message (it seems that's actually as soon as I click on anything that refreshes the page), they turn to the following:
> 
> image
> 
> That's, their `draggable` attribute is set to `true`.
> 

I confirm this, I see the same behavior

**Comment by nicolas-san** [comment]
Another thing, the account name now redirect on the settings page of the account, and when I click on the inbox (of the account, not the unified) it's open the sub-directory.
For now I can't open the root inbox of my account, only the sub folders.

**Comment by nicolas-san** [comment]
@StCyr I have PR directly to your fork, if you want to merge the prettier style fix and a little bit of css, that way you can continue on your branch and PR all at once here.

I can't spent more time on this, I have rich my limits on vue, it's new to me.

Good luck ;)

**Comment by StCyr** [comment]
Hey @ChristophWurst,

I've find out where to put the code to automaticaly expand folders when an envelope is dragged over them (see commit https://github.com/nextcloud/mail/pull/1753/commits/bf458db170dc8db96a93fd3b712852210476e6e9)

I'd like to know if you're interested in this functionality. And if so, what you think about this commit. 

For the moment the functionality is very basic: A folder will expand when a dragged element enters them. That's it. But, I can easily imagine a more complex behaviour. Examples:

- [ ] Do not expand when folder's account is not the same as envelope account
- [ ] Add a transitioning effect (some delay before expanding?)
- [ ] Auto-collapse in certain cases when leaving the drop area
- [ ] ....

What do you think about all this?

Cyrille

**Comment by StCyr** [comment]
Hi @nextcloud/mail folks,

I find my PR quite good now.

Could you please review it and give your comment?

Remaining issues:

- [ ] It seems the code isn't properly initialized: If you load the mail apps from your browser, the drag'n'drop functionality won't work directly. You first have to perform an action (eg: selecting another email than the default one displayed) before the drag'n'drop functionality will work;
- [ ] Shouldn't navigate when the moved message is not tghe one currently displayed
- [ ] Do not expand when folder's account is not the same as envelope account
- [ ] Polishing
- [ ] Problem when dragging an envelope while a new mail notification occurs (see https://github.com/nextcloud/mail/issues/2027)

Reminder, this PR should solve:

#254
#659
#670
#1538
#786

File: src/components/NavigationAccountExpandCollapse.vue [hunk 3]
```diff
@@ -48,5 +48,11 @@ export default {
 			}
 		},
 	},
+	mounted() {
+		let self = this
+		this.$el.ondragenter = function() {
```
Comment (ChristophWurst) [inline src/components/NavigationAccountExpandCollapse.vue]: pro tip: if you use an arrow function `() => {...}` `this` won't be bound to the element but the surrounding one from the Vue component

File: src/components/NavigationFolder.vue [hunk 4]
```diff
@@ -65,6 +66,12 @@ export default {
 			}
 		},
 	},
+	mounted() {
+		let self = this
+		this.$el.ondragenter = function() {
```
Comment (ChristophWurst) [inline src/components/NavigationFolder.vue]: same here

File: src/views/Home.vue [hunk 5]
```diff
@@ -84,6 +91,67 @@ export default {
...
+
+			// We don't want to use this.activeAccount here because we want it to work in the unified inbox
+			var startAccount = e.items[0].attributes.href.baseURI.match(/message\/(\d)/)
+			var targetAccount = e.droptarget.href.match(/accounts\/(\d)/)
+
+			if (targetAccount[1] == startAccount[1]) {
+				var accountId = targetAccount[1]
+
+				// We don't want to use this.activeFolder here because we want it to work in the unified inbox
+				var startFolderId = e.items[0].attributes.href.baseURI.match(/.+-([^-]+)-\d+$/)
```
Comment (ChristophWurst) [inline src/views/Home.vue]: why do we actually have to do regex matches here? Isn't there a way to transport the information in JSON and just load the data from the Vuex store?

File: src/views/Home.vue [hunk 6]
```diff
@@ -84,6 +91,67 @@ export default {
 		}
 	},
 	methods: {
+		moveMessage(e) {
+
+			// We don't want to use this.activeAccount here because we want it to work in the unified inbox
+			var startAccount = e.items[0].attributes.href.baseURI.match(/message\/(\d)/)
+			var targetAccount = e.droptarget.href.match(/accounts\/(\d)/)
+
+			if (targetAccount[1] == startAccount[1]) {
```
Comment (ChristophWurst) [inline src/views/Home.vue]: general coding tip: use *early return*s for the unlikely condition. Then the rest of the function doesn't have to be indented and the structure is a lot easier to read.

**Comment by ChristophWurst** [comment]
> It seems the code isn't properly initialized: If you load the mail apps from your browser, the drag'n'drop functionality won't work directly. You first have to perform an action (eg: selecting another email than the default one displayed) before the drag'n'drop functionality will work;

This is strange indeed. I updated the branch to the latest master and fixed a few minor things. This is still the case. I also tried updating to the latest `vue-draggable` v2.0.6 – still the same result :thinking:

**Comment by ChristophWurst** [comment]
I pushed my fixes to https://github.com/nextcloud/mail/tree/StCyr-Issue598 as I can't push to your branch. That branch also has the conflicts resolved shown here.

**Comment by jancborchardt** [comment]
> I pushed my fixes to https://github.com/nextcloud/mail/tree/StCyr-Issue598 as I can't push to your branch. That branch also has the conflicts resolved shown here.

You sure that’s correct, as this branch is called "StCyr-Issue**786**"? It also doesn’t really do anything, dragging messages e.g. from Inbox to Archive has no effect.

Or is this not ready for testing yet?

**Comment by ChristophWurst** [comment]
Now I'm confused as well. But my branch contains the dnd changes, so I guess I kinda mixed the two features together. Let me see if I can rebase now that the view source is merged.

**Comment by ChristophWurst** [comment]
Rebased

**Comment by jancborchardt** [comment]
I’m definitely on that branch (`StCyr-Issue598`, on `2277e261`), built everything, yet it does not work. I’m just trying to drag a message from e.g. the Inbox to Archive. When I drag the mail, it looks like I’m moving it, but on dropping there’s no feedback, and the mail is not moved.

What am I missing? :D

**Comment by StCyr** [comment]
> I’m definitely on that branch (`StCyr-Issue598`, on `2277e261`), built everything, yet it does not work. I’m just trying to drag a message from e.g. the Inbox to Archive. When I drag the mail, it looks like I’m moving it, but on dropping there’s no feedback, and the mail is not moved.
> 
> What am I missing? :D

Do you perfom another action before trying to drag&drop the envelope?

There's currently an issue: If you load the mail apps from your browser, the drag'n'drop functionality won't work directly. You first have to perform an action (eg: selecting another email than the default one displayed) before the drag'n'drop functionality will work;

**Comment by jancborchardt** [comment]
@StCyr could you maybe update this pull request to include @ChristophWurst’s changes? I think in the settings of your fork you can also allow contributors of the original repository to collaborate.

Alternatively maybe the easiest way is to open a new pull request directly from a branch in this repository – since you are a member of the Nextcloud organization on GitHub, you don’t need your own fork. That makes collaboration much simpler. :)

**Comment by ChristophWurst** [comment]
> @StCyr could you maybe update this pull request to include @ChristophWurst’s changes? I think in the settings of your fork you can also allow contributors of the original repository to collaborate.

You are asked when you open the PR.

> Alternatively maybe the easiest way is to open a new pull request directly from a branch in this repository – since you are a member of the Nextcloud organization on GitHub, you don’t need your own fork. That makes collaboration much simpler. :)

PRs from forks work just as fine. I could update @StCyr's other PR branches without any problems. Just this one seems to miss that permission :)

**Comment by StCyr** [comment]
The problem is that there are 2 branches "StCyr-Issue786": one in this repo and another on in my fork.

And, this PR refers to the branch in my fork.

Maybe, it's better that I create a new PR and delete the branch in my fork?

**Comment by jancborchardt** [comment]
> Maybe, it's better that I create a new PR and delete the branch in my fork?

Yeah, I’d say best delete your fork completely as your other pull request is already from a branch in this repo. And then you can create the new PR as a branch in this repository too. :)

Really looking forward to drag&drop moving too – do @-mention me and I’ll give it a test! :)

**Comment by StCyr** [comment]
Closed in favor of https://github.com/nextcloud/mail/pull/1753

**Comment by jancborchardt** [comment]
@StCyr you mean https://github.com/nextcloud/mail/pull/2834, right? :)

**Comment by StCyr** [comment]
pfff I'm tired :-p

**Pull request closed without merging**